cursor.c

```diff
@@ -380,15 +380,11 @@ int php_mongocursor_advance(mongo_cursor *cursor TSRMLS_DC)
 		}
 		/* Limit reached */
 		if (cursor->limit != 0 && cursor->at >= cursor->limit) {
-			mongo_deregister_callback_from_connection(cursor->connection, cursor);
-			php_mongo_cursor_mark_dead(cursor);
+			php_mongo_kill_cursor(cursor->connection, cursor->cursor_id TSRMLS_CC);
+			cursor->cursor_id = 0;
 			return FAILURE;
 		}
 		if (!php_mongo_get_more(cursor TSRMLS_CC)) {
-			if (cursor->connection) {
-				mongo_deregister_callback_from_connection(cursor->connection, cursor);
-			}
-			php_mongo_cursor_mark_dead(cursor);
```

Comment:
When `php_mongo_get_more()` fails (due to a socket error of some sort), it will call `php_mongo_cursor_failed()`, which in turn calls `mongo_manager_connection_deregister` and clears the cursor's connection pointer.

In that case `cursor->connection` would already be null, so we were never actually invoking `mongo_deregister_callback_from_connection()`. Beyond that, I think my commit message sufficiently explains how `php_mongo_cursor_mark_dead()` is inferior to ``php_mongo_cursor_failed()`.